proxy_worker/dispatcher.py

```diff
@@ -0,0 +1,528 @@
+
+        try:
+            grpc.channel_ready_future(channel).result(
+                timeout=self._grpc_connect_timeout)
+        except Exception as ex:
+            self._loop.call_soon_threadsafe(
+                self._grpc_connected_fut.set_exception, ex)
+            return
+        else:
+            self._loop.call_soon_threadsafe(
+                self._grpc_connected_fut.set_result, True)
+
+        stub = protos.FunctionRpcStub(channel)
+
+        def gen(resp_queue):
+            while True:
+                msg = resp_queue.get()
+                if msg is self._GRPC_STOP_RESPONSE:
+                    grpc_req_stream.cancel()
+                    return
+                yield msg
+
+        grpc_req_stream = stub.EventStream(gen(self._grpc_resp_queue))
+        try:
+            for req in grpc_req_stream:
+                self._loop.call_soon_threadsafe(
+                    self._loop.create_task, self._dispatch_grpc_request(req))
+        except Exception as ex:
+            if ex is grpc_req_stream:
+                # Yes, this is how grpc_req_stream iterator exits.
+                return
+            error_logger.exception(
+                'unhandled error in gRPC thread. Exception: {0}'.format(
+                    ''.join(traceback.format_exception(ex))))
+            raise
+
+    async def _dispatch_grpc_request(self, request):
+        content_type = request.WhichOneof("content")
+
+        match content_type:
+            case "worker_init_request":
+                request_handler = self._handle__worker_init_request
+            case "function_environment_reload_request":
+                request_handler = self._handle__function_environment_reload_request
+            case "functions_metadata_request":
+                request_handler = self._handle__functions_metadata_request
+            case "function_load_request":
+                request_handler = self._handle__function_load_request
+            case "worker_status_request":
+                request_handler = self._handle__worker_status_request
+            case "invocation_request":
+                request_handler = self._handle__invocation_request
+            case _:
+                # Don't crash on unknown messages. Log the error and return.
+                logger.error("Unknown StreamingMessage content type: %s", content_type)
+                return
+
+        resp = await request_handler(request)
+        self._grpc_resp_queue.put_nowait(resp)
+
+    async def dispatch_forever(self):  # sourcery skip: swap-if-expression
+        if DispatcherMeta.__current_dispatcher__ is not None:
+            raise RuntimeError('there can be only one running dispatcher per '
+                               'process')
+
+        self._old_task_factory = self._loop.get_task_factory()
+
+        DispatcherMeta.__current_dispatcher__ = self
+        try:
+            forever = self._loop.create_future()
+
+            self._grpc_resp_queue.put_nowait(
```

Comment:
[Checking, unaware of python data structures] If multiple thread attempt to put in the queue is that thread safe. We use something called BlockingQueue or ConcurrentQueue in Java